### DISCUSSION TITLE
TypeSetting ResNetBlock

### DISCUSSION THREAD

**Discussion by vikashg** [Project-MONAI/MONAI#3550] [Q&A]
Hi, I am trying to understand the ResNet implementation and I am struggling with this one line    
`block: Type[Union[ResNetBlock, ResNetBottleneck]]`,
[Resnet Implementation ](https://github.com/Project-MONAI/MONAI/blob/df75199b76a04d991081885888c54f8434eadba2/monai/networks/nets/resnet.py#L312)
As far as I understand the variable `block` can be of type either ResNetBlock or ResNetBottleneck. However, I am curious what does `Type` do in this context. Should it not be `NewType` or `TypeVar`. More info [here](https://stackoverflow.com/questions/58755948/what-is-the-difference-between-typevar-and-newtype)

**Comment by Nic-Ma** [comment]
Hi @vikashg ,

Thanks for your interest here and feedback, I think it's a typo, let me try to update it ASAP.

**Reply by Nic-Ma** [reply]
Hi @ericspod @wyli ,

I feel we should update the typehints `Type[]` to `TypeVar[]`, what do you think?

Thanks in advance.

**Reply by ericspod** [reply]
`Type` is [marked deprecated since Python 3.9](https://docs.python.org/3/library/typing.html#typing.Type) but we would still need to support it in earlier versions. It's replaced with using `builtins.type` as a generic type with `[]`. I don't think `TypeVar` is meant to be used in placed of `Type` but as a variable in other generic types.

**Reply by vikashg** [reply]
Hi @Nic-Ma and @ericspod, Thanks for the quick reply. A follow-up question is that am I interpreting the use of `Type` correctly here. I generally do not use the `typing` library so I am not totally familiar with it.

**Reply by Nic-Ma** [reply]
Hi @ericspod ,

Do you mean we should use `[XXX]` to replace `Type[XXX]` after python 3.9?

Thanks.

**Reply by ericspod** [reply]
It's complicated with the different versions of Python we're supporting but from 3.9 my understanding is `Type[foo]` becomes `type[foo]`, that is using the built-in `type` class. `TypeVar` is still the same, it's represents a type variable rather than a standalone type itself.

**Reply by Nic-Ma** [reply]
OK, so I think we don't need to change anything so far, right?

Thanks.

**Reply by vikashg** [reply]
It is working fine indeed. So I don't think there is any need to change.

**Reply by ericspod** [reply]
In general we'll have to update places `Type` is used eventually. In this case however I think the line should be `block: Union[ResNetBlock, ResNetBottleneck]` since `block` is an instance of one of these types rather than an instance of a type class.

**Reply by Nic-Ma** [reply]
Hi @ericspod ,

I think in this case, the `block` is a class instead of an instance, so we used the `Type`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/nets/resnet.py#L340
Right?

Thanks.

**Reply by ericspod** [reply]
Ah yes that's actually a type, so yes `Type` is what to use.
